Source/Core/VideoCommon/VertexLoaderARM64.cpp

```diff
@@ -0,0 +1,481 @@
+			ORR(scratch2_reg, scratch2_reg, scratch2_reg, ArithOption(scratch2_reg, ST_LSR, 6));
+			ORR(scratch1_reg, scratch1_reg, scratch2_reg, ArithOption(scratch2_reg, ST_LSL, 8));
+
+			// R
+			UBFM(scratch2_reg, scratch3_reg, 18, 23);
+			ORR(scratch1_reg, scratch1_reg, scratch2_reg, ArithOption(scratch2_reg, ST_LSL, 2));
+			ORR(scratch1_reg, scratch1_reg, scratch2_reg, ArithOption(scratch2_reg, ST_LSR, 4));
+
+			STR(INDEX_UNSIGNED, scratch1_reg, dst_reg, m_dst_ofs);
+			load_bytes = 3;
+			break;
+	}
+	if (attribute == DIRECT)
+		m_src_ofs += load_bytes;
+}
+
+void VertexLoaderARM64::GenerateVertexLoader()
+{
+	// R0 - Source pointer
+	// R1 - Destination pointer
+	// R2 - Count
+	// R30 - LR
+	//
+	// R0 return how many
+	//
+	// Registers we don't have to worry about saving
+	// R9-R17 are caller saved temporaries
+	// R18 is a temporary or platform specific register(iOS)
+	//
+	// VFP registers
+	// We can touch all except v8-v15
+	// If we need to use those, we need to retain the lower 64bits(!) of the register
+
+	MOV(skipped_reg, WSP);
+	MOV(saved_count, count_reg);
+
+	const u8* loop_start = GetCodePtr();
+
+	if (m_VtxDesc.PosMatIdx)
+	{
+		LDRB(INDEX_UNSIGNED, scratch1_reg, src_reg, m_src_ofs);
+		AND(scratch1_reg, scratch1_reg, 0, 5);
+		STR(INDEX_UNSIGNED, scratch1_reg, dst_reg, m_dst_ofs);
+		m_native_components |= VB_HAS_POSMTXIDX;
+		m_native_vtx_decl.posmtx.components = 4;
+		m_native_vtx_decl.posmtx.enable = true;
+		m_native_vtx_decl.posmtx.offset = m_dst_ofs;
+		m_native_vtx_decl.posmtx.type = VAR_UNSIGNED_BYTE;
+		m_native_vtx_decl.posmtx.integer = true;
+		m_src_ofs += sizeof(u8);
+		m_dst_ofs += sizeof(u32);
+	}
+
+	u32 texmatidx_ofs[8];
+	const u64 tm[8] = {
+		m_VtxDesc.Tex0MatIdx, m_VtxDesc.Tex1MatIdx, m_VtxDesc.Tex2MatIdx, m_VtxDesc.Tex3MatIdx,
+		m_VtxDesc.Tex4MatIdx, m_VtxDesc.Tex5MatIdx, m_VtxDesc.Tex6MatIdx, m_VtxDesc.Tex7MatIdx,
+	};
+	for (int i = 0; i < 8; i++)
+	{
+		if (tm[i])
+			texmatidx_ofs[i] = m_src_ofs++;
+	}
+
+	GetVertexAddr(ARRAY_POSITION, m_VtxDesc.Position, EncodeRegTo64(scratch1_reg));
+	ReadVertex(m_VtxDesc.Position, m_VtxAttr.PosFormat, m_VtxAttr.PosElements + 2, 3,
+	           m_VtxAttr.ByteDequant, m_VtxAttr.PosFrac, &m_native_vtx_decl.position);
+
+	if (m_VtxDesc.Normal)
+	{
+		static const u8 map[8] = {7, 6, 15, 14};
+		u8 scaling_exponent = map[m_VtxAttr.NormalFormat];
+
+		for (int i = 0; i < (m_VtxAttr.NormalElements ? 3 : 1); i++)
+		{
+			if (!i || m_VtxAttr.NormalIndex3)
+			{
+				GetVertexAddr(ARRAY_NORMAL, m_VtxDesc.Normal, EncodeRegTo64(scratch1_reg));
+				int elem_size = 1 << (m_VtxAttr.NormalFormat / 2);
+				ADD(EncodeRegTo64(scratch1_reg), EncodeRegTo64(scratch1_reg), i * elem_size * 3);
+			}
+			int bytes_read = ReadVertex(m_VtxDesc.Normal, m_VtxAttr.NormalFormat, 3, 3,
+			                            true, scaling_exponent, &m_native_vtx_decl.normals[i]);
+
+			ADD(EncodeRegTo64(scratch1_reg), EncodeRegTo64(scratch1_reg), bytes_read);
```

Comment:
what's the purpose of continually adding to the scratch register; can't you just keep a running offset and use offsetted loads for reading from the input?